x/superfluid/keeper/stake_test.go

```diff
@@ -1028,6 +1031,400 @@ func (s *KeeperTestSuite) TestRefreshIntermediaryDelegationAmounts() {
+			s.SetupTest()
+			s.Ctx = s.Ctx.WithBlockTime(defaultJoinTime)
+			// We bundle all migration setup into a single function to avoid repeating the same code for each test case.
+			_, _, lock, _, _, _, _, originalValAddr := s.SetupUnbondConvertAndStakeTest(s.Ctx, true, tc.superfluidUndelegating, false, false)
+
+			// testing params
+			sender := sdk.MustAccAddressFromBech32(lock.Owner)
+			valAddr := s.SetupValidator(stakingtypes.Bonded)
+			minAmountToStake := sdk.ZeroInt()
+
+			balanceBeforeConvertLockToStake := s.App.BankKeeper.GetAllBalances(s.Ctx, sender)
+
+			// system under test
+			totalAmtConverted, sharesDelegated, err := s.App.SuperfluidKeeper.UnbondConvertAndStake(s.Ctx, lock.ID, sender.String(), valAddr.String(), minAmountToStake)
+			if tc.expectedError {
+				s.Require().Error(err)
+				return
+			}
+			s.Require().NoError(err)
+
+			// Staking & Delegation check
+			// check if old delegation is succesfully deleted
+			_, found := s.App.StakingKeeper.GetDelegation(s.Ctx, sender, originalValAddr)
+			s.Require().False(found)
+			// check if delegation amount matches
+			delegation, found := s.App.StakingKeeper.GetDelegation(s.Ctx, sender, valAddr)
+			s.Require().True(found)
+			s.Require().Equal(delegation.Shares, sharesDelegated)
+			s.Require().True(totalAmtConverted.ToDec().Equal(delegation.Shares))
+
+			// Superfluid check
+			// The synthetic lockup should be deleted.
+			_, err = s.App.LockupKeeper.GetSyntheticLockup(s.Ctx, lock.ID, keeper.StakingSyntheticDenom(lock.Coins[0].Denom, valAddr.String()))
+			s.Require().Error(err)
+
+			// intermediary account should have been deleted in all case
+			if lock.ID != 0 {
+				_, err = s.App.LockupKeeper.GetSyntheticLockup(s.Ctx, lock.ID, keeper.UnstakingSyntheticDenom(lock.Coins[0].Denom, valAddr.String()))
+				s.Require().Error(err)
+			}
+
+			// Lock check
+			_, err = s.App.LockupKeeper.GetLockByID(s.Ctx, lock.ID)
+			s.Require().Error(err)
+
+			// Bank check
+			balanceAfterConvertLockToStake := s.App.BankKeeper.GetAllBalances(s.Ctx, sender)
+			s.Require().True(balanceBeforeConvertLockToStake.IsEqual(balanceAfterConvertLockToStake))
+		})
+	}
+}
+
+func (s *KeeperTestSuite) TestConvertLockToStake() {
+	defaultJoinTime := s.Ctx.BlockTime()
+	type tc struct {
+		superfluidUndelegating bool
+		unlocking              bool
+
+		useMinAmountToStake    bool
+		senderIsNotOwnerOfLock bool
+
+		expectedError bool
+	}
+	testCases := map[string]tc{
+		"lock that is superfluid delegated": {},
+		"lock that is superfluid undelegating": {
+			unlocking:              true,
+			superfluidUndelegating: true,
+		},
+		"lock that is unlocking": {
+			unlocking:              true,
+			superfluidUndelegating: false,
+		},
+		// error cases
+		"error: min amount to stake greater than actual amount": {
+			useMinAmountToStake: true,
+			expectedError:       true,
+		},
+		"error: sender is not owner of lock ": {
+			senderIsNotOwnerOfLock: true,
+			expectedError:          true,
+		},
```

Comment:
How are we handling locks migrated to CL? Can we add tests for them as well?

Reply:
ditto for above